### DISCUSSION TITLE
How to implement (and use) custom themes for `std-widgets.slint`?

### DISCUSSION THREAD

**Discussion by Reverier-Xu** [slint-ui/slint#7232] [Q&A] [answered]
Hi, I'm trying to use slint to develop application with my own design.

In Qt/QML, we could change the outlook of QtQuick Controls by deriving [QtQuick.Templates](https://doc.qt.io/qt-6/qtquicktemplates2-index.html), then provide a style plugin as a library, I have developed [a set of styled controls](https://github.com/Reverier-Xu/BitWave/blob/main/lib/rxui/src/CMakeLists.txt#L30) in my media player, and I want to do the same thing to styling `std-widgets.slint` in slint, but it seems do not have similar features like QtQuick.Templates?

I also tried write my widgets from scratch using basic elements, but I found [`ListView`](https://docs.slint.dev/latest/docs/slint/reference/std-widgets/views/listview/) has "Elements are only instantiated if they are visible", which I can't find the implementation in the [source code of listview.slint](https://github.com/slint-ui/slint/blob/master/internal/compiler/widgets/common/listview.slint). Although I can implement the display optimizations myself, I still want to simply styling `std-widgets.slint` or reuse logic infrastructures instead of re-implement wheels. 

Is there some way to implement themes like `fluent` or `cosmic` without monkey-editing slint's source code?

**Comment by Reverier-Xu** [comment]
Additional: [`ListView`](https://doc.qt.io/qt-6/qml-qtquick-listview.html) in QtQuick is provided as a basic element, not a styled-control.

Maybe we could do the same thing? provide something like `ListView` as a basic element and styling & extending it as another name in `std-widgets.slint`.

**Comment by ogoffart** [comment, marked as answer]
> Is there some way to implement themes like fluent or cosmic without monkey-editing slint's source code?

The way this was designed is that one would adapt the include path so that "std-widgets.slint" point to your own. 
But in practice this is hardly possible because the styles now rely on some internal that might not even be accessible.
This is tracking in issue https://github.com/slint-ui/slint/issues/45 , but we are currently not really working on that.

Regarding the ListView question, the ListView has special handling [in the compiler](https://github.com/slint-ui/slint/blob/87702e9b55af1a003d01de8bdd57d92d706247d5/internal/compiler/object_tree.rs#L1593) so that at runtime, a special layout function is called: https://github.com/slint-ui/slint/blob/87702e9b55af1a003d01de8bdd57d92d706247d5/internal/core/model.rs#L959

This works by just matching the name of the component to "ListView", so if you declare a component called ListView, it will have the same behavior (https://github.com/slint-ui/slint/issues/861)
